### PR TITLE
feat(query): quantile_over_time support

### PR DESCRIPTION
**Pull request by HimaVarsha94**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Prometheus quantile_over_time function support

### CODE REVIEW

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 1]
```diff
@@ -584,4 +584,81 @@ class ChangesChunkedFunctionL extends ChangesChunkedFunction with
     }
     changes += longReader.changes(longVect, startRowNum, endRowNum)
   }
-}
\ No newline at end of file
+}
+
+abstract class QuantileOverTimeChunkedFunction(funcParams: Seq[Any], var result: Double = Double.NaN
```
[the remark this answers is not in the document]
Comment (HimaVarsha94) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: quantile refers to the value between 0 to 1. It is not the output of quantile over time, so I cannot use it. I will use quantileResult instead

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 2]
```diff
@@ -584,4 +584,81 @@ class ChangesChunkedFunctionL extends ChangesChunkedFunction with
...
+          counter += 1
+        }
+        rowNum += 1
+      }
+      val sortedValues = values.sorted
+      val rank = q*(counter - 1)
+      val lowerIndex = Math.max(0, Math.floor(rank))
+      val upperIndex = Math.min(counter - 1, lowerIndex + 1)
+      val weight = rank - math.floor(rank)
+      sortedValues(lowerIndex.toInt)*(1-weight) + sortedValues(upperIndex.toInt)*weight
```
[the remark this answers is not in the document]
Comment (HimaVarsha94) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: I do not think it will be able to access the element. Good point. Thanks!

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 3]
```diff
@@ -584,4 +586,89 @@ class ChangesChunkedFunctionL extends ChangesChunkedFunction with
     }
     changes += longReader.changes(longVect, startRowNum, endRowNum)
   }
-}
\ No newline at end of file
+}
+
+abstract class QuantileOverTimeChunkedFunction(funcParams: Seq[Any], var quantileResult: Double = Double.NaN
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Minor: I would put `var quantileResult` on a separate line.

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 4]
```diff
@@ -584,4 +586,89 @@ class ChangesChunkedFunctionL extends ChangesChunkedFunction with
...
+                                endRowNum: Int): Unit = {
+    require(funcParams.head.isInstanceOf[Number], "quantile parameter must be a number")
+    val q = funcParams.head.asInstanceOf[Number].doubleValue()
+    var counter = 0
+    quantileResult = if (q < 0) Double.NegativeInfinity
+    else if (q > 1) Double.PositiveInfinity
+    else {
+      var rowNum = startRowNum
+      val it = doubleReader.iterate(doubleVect, startRowNum)
+      var values = new ArrayBuffer[Double]
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Unfortunately `ArrayBuffer[Double]` will box the doubles, causing a Java object to be wrapped around each number, and using a ton of memory.  Instead, use `debox.Buffer` which will not box and is much more efficient.

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 5]
```diff
@@ -584,4 +586,89 @@ class ChangesChunkedFunctionL extends ChangesChunkedFunction with
...
+    quantileResult = if (q < 0) Double.NegativeInfinity
+    else if (q > 1) Double.PositiveInfinity
+    else {
+      var rowNum = startRowNum
+      val it = longReader.iterate(longVect, startRowNum)
+      var values = new ArrayBuffer[Double]
+      while (rowNum <= endRowNum) {
+        var nextvalue = it.next
+        // There are many possible values of NaN.  Use a function to ignore them reliably.
+        if (!JLDouble.isNaN(nextvalue)) {
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: This check can be skipped.  Longs are never going to be `NaN`

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 6]
```diff
@@ -584,4 +586,89 @@ class ChangesChunkedFunctionL extends ChangesChunkedFunction with
...
+          values += nextvalue
+          counter += 1
+        }
+        rowNum += 1
+      }
+      val sortedValues = values.sorted
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: This part of the algorithm which calculates the rank can be abstracted out.

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 7]
```diff
@@ -584,4 +586,89 @@ class ChangesChunkedFunctionL extends ChangesChunkedFunction with
...
+                              endRowNum: Int): Unit = {
+    var counter = 0
+    require(funcParams.head.isInstanceOf[Number], "quantile parameter must be a number")
+    val q = funcParams.head.asInstanceOf[Number].doubleValue()
+    quantileResult = if (q < 0) Double.NegativeInfinity
+    else if (q > 1) Double.PositiveInfinity
+    else {
+      var rowNum = startRowNum
+      val it = longReader.iterate(longVect, startRowNum)
+      var values = new ArrayBuffer[Double]
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Again, `debox.Buffer[Double]`.

File: query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala [hunk 8]
```diff
@@ -298,18 +298,19 @@ object RangeFunction {
                           func: Option[InternalRangeFunction],
                           funcParams: Seq[Any] = Nil): RangeFunctionGenerator = {
     func match {
-      case None                 => () => new LastSampleChunkedFunctionL
-      case Some(CountOverTime)  => () => new CountOverTimeChunkedFunction()
-      case Some(SumOverTime)    => () => new SumOverTimeChunkedFunctionL
+      case None                   => () => new LastSampleChunkedFunctionL
+      case Some(CountOverTime)    => () => new CountOverTimeChunkedFunction()
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala]: Why are these moved?

Comment (HimaVarsha94) [reply query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala]: Because I was asked to align all of them

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 9]
```diff
@@ -584,4 +586,89 @@ class ChangesChunkedFunctionL extends ChangesChunkedFunction with
...
+}
+
+abstract class QuantileOverTimeChunkedFunction(funcParams: Seq[Any],
+                                               var quantileResult: Double = Double.NaN)
+  extends ChunkedRangeFunction[TransientRow] {
+  override final def reset(): Unit = { quantileResult = Double.NaN }
+  final def apply(endTimestamp: Long, sampleToEmit: TransientRow): Unit = {
+    sampleToEmit.setValues(endTimestamp, quantileResult)
+  }
+  def calculateRank(q: Double, counter: Long): (Double, Double, Double) = {
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Shouldn't indexes be Int's, not Doubles?

Comment (HimaVarsha94) [reply query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala]: Yes, it should be. I typecast them later. Since rank is a double, when I calculate it returns a Double.
